src/utils/futures_util/src/pausable.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use futures::Stream;
+use pin_project_lite::pin_project;
+
+pin_project! {
+    #[derive(Debug)]
+    #[must_use = "streams do nothing unless polled"]
+    pub struct Pausable<St>
+        where St: Stream
+    {
+        #[pin]
+        stream: St,
+        paused: Arc<AtomicBool>,
+    }
+}
+
+#[derive(Clone)]
+pub struct Valve {
+    paused: Arc<AtomicBool>,
+}
+
+impl Valve {
+    /// Pause the stream controlled by the valve.
+    pub fn pause(&self) {
+        self.paused.store(true, Ordering::Relaxed)
+    }
+
+    /// Resume the stream controlled by the valve.
+    pub fn resume(&self) {
+        self.paused.store(false, Ordering::Relaxed);
+    }
+}
+
+impl<St> Pausable<St>
+where
+    St: Stream,
+{
+    pub(crate) fn new(stream: St) -> (Self, Valve) {
+        let paused = Arc::new(AtomicBool::new(false));
+        (
+            Pausable {
+                stream,
+                paused: paused.clone(),
+            },
+            Valve { paused },
+        )
+    }
+}
+
+impl<St> Stream for Pausable<St>
+where
+    St: Stream,
+{
+    type Item = St::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        if this.paused.load(Ordering::Relaxed) {
+            Poll::Pending
```

Comment:
This could potentially lead to deadlock.

Reply:
Can you give me more explanation? I can only find https://docs.rs/futures-util/0.3.30/src/futures_util/stream/pending.rs.html#32

Reply:
Found the article, but I can't find the solution.

Reply:
Oh, should I register the paused as a signal?

Reply:
Here's a simple reproduction. 🥹

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=79b7a4ce0464059b0390ee56c01ee625

Reply:
> Oh, should I register the paused as a signal?

Yeah. I guess using anything else with a wait-queue mechanism should work. Although I guess the deadlock won't occur with the usage of this PR, but since it's made a general utility, we should be more careful about it.